src/util/circular_buffer.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+
+#include "redismodule.h"
+#include "circular_buffer.h"
+#include "src/rmalloc.h"
+#include "src/util/likely.h"
+
+#include <stdatomic.h>
+
+// Circular buffer structure.
+// The buffer is of fixed size.
+// Items are removed by order of insertion, similar to a queue.
+struct _CircularBuffer {
+	char *read;                   // read data from here
+	_Atomic uint64_t write;       // write offset into data
+	size_t item_size;             // item size in bytes
+	_Atomic uint64_t item_count;  // current number of items in buffer
+	uint64_t item_cap;            // max number of items held by buffer
```

Comment:
The comment here with variable name is a bit misleading.
I thought it was the max capacity of the container.
Consider changing variable name or comment.

Reply:
Not sure I follow - the `item_cap` is indeed the max capacity of the buffer.
Please be more specific.

Reply:
In that case here is the suggested change:
```suggestion
	uint64_t item_cap;            // max number of items that can be held in buffer
```